pkg/resource/addon/hooks.go

```diff
@@ -13,6 +13,223 @@
+)
+
+// Taken from the list of addon statuses in the EKS API documentation:
+// https://docs.aws.amazon.com/eks/latest/APIReference/API_Addon.html#AmazonEKS-Type-Addon-status
+const (
+	StatusActive       = "ACTIVE"
+	StatusCreating     = "CREATING"
+	StatusCreateFailed = "CREATE_FAILED"
+	StatusUpdating     = "UPDATING"
+	StatusUpdateFailed = "UPDATE_FAILED"
+	StatusDeleting     = "DELETING"
+	StatusDeleteFailed = "DELETE_FAILED"
+	StatusDegraded     = "DEGRADED"
+)
+
+var (
+	requeueWaitWhileDeleting = ackrequeue.NeededAfter(
+		fmt.Errorf("addon in '%s' state, cannot be modified or deleted", StatusDeleting),
+		ackrequeue.DefaultRequeueAfterDuration,
+	)
+)
+
+var (
+	// TerminalStatuses defines the list of statuses that are terminal for an addon
+	TerminalStatuses = []string{
+		StatusCreateFailed,
+		StatusUpdateFailed,
+		StatusDeleteFailed,
+		// Still not sure if we should consider DEGRADED as terminal
+		// StatusDegraded,
+	}
+)
+
+// addonActive returns true if the supplied addib is in an active state
+func addonActive(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	cs := *r.ko.Status.Status
+	return cs == StatusActive
+}
+
+// addonCreating returns true if the supplied addon is in a creating state
+func addonCreating(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	cs := *r.ko.Status.Status
+	return cs == StatusCreating
+}
+
+// addonDeleting returns true if the supplied addon is in a deleting state
+func addonDeleting(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	cs := *r.ko.Status.Status
+	return cs == StatusDeleting
+}
+
+// addonHasTerminalStatus returns true if the supplied addon is in a terminal state
+func addonHasTerminalStatus(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	cs := *r.ko.Status.Status
+	for _, ts := range TerminalStatuses {
+		if cs == ts {
+			return true
+		}
+	}
+	return false
+}
+
+// requeueWaitUntilCanModify returns a `ackrequeue.RequeueNeededAfter` struct
+// explaining the addon cannot be modified until it reaches an active status.
+func requeueWaitUntilCanModify(r *resource) *ackrequeue.RequeueNeededAfter {
+	if r.ko.Status.Status == nil {
+		return nil
+	}
+	status := *r.ko.Status.Status
+	return ackrequeue.NeededAfter(
+		fmt.Errorf("addon in '%s' state, cannot be modified until '%s'",
+			status, StatusActive),
+		ackrequeue.DefaultRequeueAfterDuration,
+	)
+}
+
+// returnAddonUpdating will set synced to false on the resource and
+// return an async requeue error to signify that the resource should be
+// forcefully requeued in order to pick up the 'UPDATING' status.
+func returnAddonUpdating(r *resource) (*resource, error) {
+	msg := "Addon is currently being updated"
+	ackcondition.SetSynced(r, corev1.ConditionFalse, &msg, nil)
+	return r, ackrequeue.NeededAfter(
+		fmt.Errorf("addon in '%s' state, cannot be modified until '%s'",
+			StatusUpdating, StatusActive),
+		15,
+	)
+}
```

Comment:
most of these functions can be easily generated... probably something to attack first before going after the the `sdkUpdate` generation